File: LLM_experiments/java/results_claude-opus-4/solution_try_3/src/main/java/com/phylogenetic/tree/TreeBuilder.java
package com.phylogenetic.tree;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Builds a phylogenetic tree using single-linkage clustering.
 */
public class TreeBuilder {
  private static final Logger logger = LoggerFactory.getLogger(TreeBuilder.class);
  
  /**
   * Build a phylogenetic tree from similarity scores.
   */
  public PhylogeneticTree buildTree(Map<String, Map<String, Integer>> scoreMatrix) {
    logger.info("Building phylogenetic tree from {} species", scoreMatrix.size());
    
    // Initialize clusters - each species starts as its own cluster
    Map<String, PhylogeneticNode> clusters = new HashMap<>();
    for (String species : scoreMatrix.keySet()) {
      clusters.put(species, new PhylogeneticNode(species));
    }
    
    // Track min and max scores for validation
    int minScore = Integer.MAX_VALUE;
    int maxScore = Integer.MIN_VALUE;
    
    // Build similarity pairs for efficient processing
    List<SimilarityPair> pairs = buildSimilarityPairs(scoreMatrix);
    for (SimilarityPair pair : pairs) {
      minScore = Math.min(minScore, pair.score);
      maxScore = Math.max(maxScore, pair.score);
    }
    
    // Sort pairs by score (descending - highest similarity first)
    pairs.sort((a, b) -> Integer.compare(b.score, a.score));
    
    // Perform hierarchical clustering
    int mergeCount = 0;
    for (SimilarityPair pair : pairs) {
      if (clusters.size() == 1) {
        break; // Tree is complete
      }
      
      // Check if both clusters still exist (haven't been merged)
      if (clusters.containsKey(pair.species1) && clusters.containsKey(pair.species2)) {
        // Create new internal node
        PhylogeneticNode left = clusters.get(pair.species1);
        PhylogeneticNode right = clusters.get(pair.species2);
        
        String newNodeName = String.format("Node_%d", ++mergeCount);
        PhylogeneticNode newNode = new PhylogeneticNode(
            newNodeName, left, right, pair.score);
        
        logger.debug("Merging {} and {} at similarity score {}", 
            pair.species1, pair.species2, pair.score);
        
        // Remove old clusters
        clusters.remove(pair.species1);
        clusters.remove(pair.species2);
        
        // Add new cluster with combined name for further merging
        String combinedKey = pair.species1 + "_" + pair.species2;
        clusters.put(combinedKey, newNode);
        
        // Update score matrix for the new cluster
        updateScoreMatrix(scoreMatrix, pair.species1, pair.species2, 
            combinedKey, newNode);
      }
    }
    
    // The last remaining cluster is the root
    // if (clusters.size() != 1) {
    //   throw new IllegalStateException(
    //       "Tree construction failed. Remaining clusters: " + clusters.size());
    // }
    
    PhylogeneticNode root = clusters.values().iterator().next();
    logger.info("Tree construction complete. Root height: {}", root.getHeight());
    
    return new PhylogeneticTree(root, minScore, maxScore);
  }
  
  /**
   * Build list of all similarity pairs from the score matrix.
   */
  private List<SimilarityPair> buildSimilarityPairs(
      Map<String, Map<String, Integer>> scoreMatrix) {
    List<SimilarityPair> pairs = new ArrayList<>();
    Set<String> processed = new HashSet<>();
    
    for (Map.Entry<String, Map<String, Integer>> entry : scoreMatrix.entrySet()) {
      String species1 = entry.getKey();
      processed.add(species1);
      
      for (Map.Entry<String, Integer> innerEntry : entry.getValue().entrySet()) {
        String species2 = innerEntry.getKey();
        
        // Skip self-comparisons and already processed pairs
        if (!species1.equals(species2) && !processed.contains(species2)) {
          pairs.add(new SimilarityPair(species1, species2, innerEntry.getValue()));
        }
      }
    }
    
    return pairs;
  }
  
  /**
   * Update score matrix after merging two clusters.
   * Uses single-linkage: take maximum similarity to any member.
   */
  private void updateScoreMatrix(Map<String, Map<String, Integer>> scoreMatrix,
      String cluster1, String cluster2, String newClusterKey, 
      PhylogeneticNode newNode) {
    
    Map<String, Integer> newScores = new HashMap<>();
    
    // Get all species in the new cluster
    Set<String> mergedSpecies = new HashSet<>(
        newNode.getLeafNames());
    
    // Calculate scores between new cluster and all other entities
    for (String entity : scoreMatrix.keySet()) {
      if (!entity.equals(cluster1) && !entity.equals(cluster2)) {
        int maxScore = Integer.MIN_VALUE;
        
        // Find maximum score between any species in the clusters
        for (String speciesInNew : mergedSpecies) {
          if (scoreMatrix.containsKey(speciesInNew)) {
            Integer score = scoreMatrix.get(speciesInNew).get(entity);
            if (score == null && scoreMatrix.containsKey(entity)) {
              score = scoreMatrix.get(entity).get(speciesInNew);
            }
            if (score != null) {
              maxScore = Math.max(maxScore, score);
            }
          }
        }
        
        if (maxScore > Integer.MIN_VALUE) {
          newScores.put(entity, maxScore);
          // Update reverse direction
          scoreMatrix.get(entity).put(newClusterKey, maxScore);
        }
      }
    }
    
    scoreMatrix.put(newClusterKey, newScores);
  }
  
  /**
   * Helper class to store similarity pairs.
   */
  private static class SimilarityPair {
    final String species1;
    final String species2;
    final int score;
    
    SimilarityPair(String species1, String species2, int score) {
      this.species1 = species1;
      this.species2 = species2;
      this.score = score;
    }
  }
}

File: LLM_experiments/python/results_claude-opus-4/solution_try_1/src/tree_builder.py
"""
Module for building phylogenetic trees using single-linkage clustering.
"""
import logging
from typing import Dict, Tuple, List, Set
import numpy as np

from .tree_structures import TreeNode, PhylogeneticTree
from .score_saver import ScoreSaver


class SingleLinkageTreeBuilder:
    """Builds phylogenetic trees using single-linkage clustering with similarity scores."""
    
    def __init__(self):
        """Initialize the tree builder."""
        self.logger = logging.getLogger(__name__)
    
    def build_tree(self, scores: Dict[Tuple[str, str], int]) -> PhylogeneticTree:
        """
        Build a phylogenetic tree using single-linkage clustering.
        
        Args:
            scores: Dictionary mapping species pairs to similarity scores
            
        Returns:
            PhylogeneticTree object
        """
        # Extract unique species names
        species = set()
        for (sp1, sp2) in scores.keys():
            species.add(sp1)
            species.add(sp2)
        
        species_list = sorted(list(species))
        self.logger.info(f"Building tree for {len(species_list)} species")
        
        # Initialize clusters - each species starts as its own cluster
        clusters = {}
        for sp in species_list:
            node = TreeNode(name=sp, height=0.0)
            clusters[sp] = node
        
        # Track which species are in which cluster
        cluster_members = {sp: {sp} for sp in species_list}
        
        # Build tree by iteratively merging most similar clusters
        merge_count = 0
        while len(clusters) > 1:
            # Find the pair of clusters with highest similarity
            max_similarity = float('-inf')
            best_pair = None
            
            cluster_names = list(clusters.keys())
            for i in range(len(cluster_names)):
                for j in range(i + 1, len(cluster_names)):
                    cluster1 = cluster_names[i]
                    cluster2 = cluster_names[j]
                    
                    # Single linkage: find maximum similarity between any pair
                    similarity = self._get_cluster_similarity(
                        cluster_members[cluster1],
                        cluster_members[cluster2],
                        scores
                    )
                    
                    if similarity > max_similarity:
                        max_similarity = similarity
                        best_pair = (cluster1, cluster2)
            
            if best_pair is None:
                break
            
            # Merge the two most similar clusters
            cluster1_name, cluster2_name = best_pair
            merge_count += 1
            
            self.logger.debug(
                f"Merge {merge_count}: {cluster1_name} + {cluster2_name} "
                f"at similarity {max_similarity}"
            )
            
            # Create new internal node
            new_node = TreeNode(height=float(max_similarity))
            new_node.add_child(clusters[cluster1_name])
            new_node.add_child(clusters[cluster2_name])
            
            # Update clusters
            new_cluster_name = f"cluster_{merge_count}"
            clusters[new_cluster_name] = new_node
            
            # Update cluster members
            cluster_members[new_cluster_name] = (
                cluster_members[cluster1_name] | cluster_members[cluster2_name]
            )
            
            # Remove merged clusters
            del clusters[cluster1_name]
            del clusters[cluster2_name]
            del cluster_members[cluster1_name]
            del cluster_members[cluster2_name]
        
        # The last remaining cluster is the root
        root = list(clusters.values())[0] if clusters else None
        
        tree = PhylogeneticTree(root=root)
        self.logger.info(f"Tree construction complete with {merge_count} merges")
        
        return tree
    
    def _get_cluster_similarity(
        self, 
        members1: Set[str], 
        members2: Set[str], 
        scores: Dict[Tuple[str, str], int]
    ) -> int:
        """
        Calculate single-linkage similarity between two clusters.
        
        Returns the maximum similarity between any pair of species
        from the two clusters.
        """
        max_similarity = float('-inf')
        
        for sp1 in members1:
            for sp2 in members2:
                # Handle both orderings of the tuple key
                if (sp1, sp2) in scores:
                    similarity = scores[(sp1, sp2)]
                elif (sp2, sp1) in scores:
                    similarity = scores[(sp2, sp1)]
                else:
                    continue
                
                max_similarity = max(max_similarity, similarity)
        
        return max_similarity
    
    def build_from_file(self, blosum_type: int) -> PhylogeneticTree:
        """
        Build tree from saved similarity scores file.
        
        Args:
            blosum_type: BLOSUM matrix type (50 or 62)
            
        Returns:
            PhylogeneticTree object
        """
        # Load scores from file
        saved_scores = ScoreSaver.load_scores(blosum_type)
        
        # Convert back to tuple format
        scores = {}
        for key, value in saved_scores.items():
            species1, species2 = key.split('_', 1)
            scores[(species1, species2)] = value
        
        return self.build_tree(scores)